target/linux/realtek/dts-5.10/rtl8382_panasonic_m16eg-pn28160k.dts

```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "rtl838x.dtsi"
+#include "rtl83xx_panasonic_mxxeg-pn28xx0k.dtsi"
+
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {
+	compatible = "panasonic,m16eg-pn28160k", "realtek,rtl8382-soc";
+	model = "Panasonic Switch-M16eG PN28160K";
+
+	aliases {
+		led-boot = &led_status_eco_green;
+		led-failsafe = &led_status_eco_amber;
+		led-running = &led_status_eco_green;
+		led-upgrade = &led_status_eco_green;
+	};
+
+	/*
+	 * sfp0/1 are "combo" port with each TP port (23/24), and they are
+	 * connected to the RTL8218FB. Currently, there is no support for
+	 * the chip and only TP ports work by the RTL8218D support.
+	 */
+	sfp0: sfp-p23 {
+		compatible = "sff,sfp";
+		i2c-bus = <&i2c0>;
+		tx-fault-gpio = <&gpio1 0 GPIO_ACTIVE_HIGH>;
+		tx-disable-gpio = <&gpio1 1 GPIO_ACTIVE_HIGH>;
+		mod-def0-gpio = <&gpio1 2 GPIO_ACTIVE_LOW>;
+		los-gpio = <&gpio1 3 GPIO_ACTIVE_HIGH>;
+	};
+
+	sfp1: sfp-p24 {
+		compatible = "sff,sfp";
+		i2c-bus = <&i2c1>;
+		tx-fault-gpio = <&gpio1 4 GPIO_ACTIVE_HIGH>;
+		tx-disable-gpio = <&gpio1 5 GPIO_ACTIVE_HIGH>;
+		mod-def0-gpio = <&gpio1 6 GPIO_ACTIVE_LOW>;
+		los-gpio = <&gpio1 7 GPIO_ACTIVE_HIGH>;
+	};
+};
+
+&leds {
+	led_status_eco_amber: led-5 {
+		label = "amber:status_eco";
+		gpios = <&gpio2 1 GPIO_ACTIVE_LOW>;
+		color = <LED_COLOR_ID_AMBER>;
+		function = LED_FUNCTION_STATUS;
+		function-enumerator = <1>;
+	};
+
+	led_status_eco_green: led-6 {
+		label = "green:status_eco";
+		gpios = <&gpio2 2 GPIO_ACTIVE_LOW>;
+		color = <LED_COLOR_ID_GREEN>;
+		function = LED_FUNCTION_STATUS;
+		function-enumerator = <2>;
+	};
+};
+
+&i2c_gpio_0 {
+	scl-gpios = <&gpio0 0 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
+	sda-gpios = <&gpio0 1 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
+};
+
+&i2c_gpio_1 {
+	scl-gpios = <&gpio0 12 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
+	sda-gpios = <&gpio0 13 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
+};
+
+&gpio2 {
+	interrupt-controller;
+	#interrupt-cells = <2>;
+	interrupt-parent = <&gpio0>;
+	interrupts = <2 IRQ_TYPE_EDGE_FALLING>;
+
+	/* GPIO12 (IO1_4): RTL8218FB */
+	ext_switch_reset {
+		gpio-hog;
+		gpios = <12 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "ext-switch-reset";
```

Comment:
If this is the reset line for the RTL8218FB phy, shouldn't that go into `reset-gpios` on the MDIO (device) node?

Reply:
Indeed, it seems to be better than gpio-hog in this case.

Reply:
I tried to use of ```reset-gpios``` instead, but it breaks networking on Switch-M48eG PN28480K. The GPIO pin for resetting of phys will be requested before probing I2C and PCA9539 chip, then, the kernel shows some errors and networking will be broken.

```
[    5.632936] Probing RTL838X eth device pdev: 82111a00, dev: 82111a10
[    5.746637] Found SoC ID: 8393: RTL8393, family 8390
[    5.812299] Using MAC 0000005040101ab6
[    5.866193] mdio_bus mdio-bus: mii_bus mdio-bus couldn't get reset GPIO
[    5.954230] Error setting up netdev, freeing it again.
...
[    6.324806] pca953x 0-0075: using no AI
...
[    6.637545] RTL839X Chip-Info: a0016290, version B
[    6.701197] Probing RTL838X eth device pdev: 82111a00, dev: 82111a10
[    6.813966] Found SoC ID: 8393: RTL8393, family 8390
[    6.879439] Using MAC 0000005040101ab6
[    6.933084] mdio_bus mdio-bus: MDIO device at address 0 is missing.
[    8.270006] RTL839X Chip-Info: a0016290, version B
...
[   18.573278] In rtl83xx_vlan_setup
[   18.616963] UNKNOWN_MC_PMASK: 001fffffffffffff
[   18.675507] VLAN profile 0: L2 learning: 1, UNKN L2MC FLD PMSK 511,          UNKN IPMC FLD PMSK 511, UNKN IPv6MC FLD PMSK: 511
[   18.675518] VLAN profile 0: raw 000001ff, 003fe3ff
[   19.935092] rtl83xx_enable_phy_polling:     ffffffffffff
[   21.056181] rtl83xx-switch switch@1b000000 lan1 (uninitialized): PHY [rtl838x slave mii-0:00] driver [Realtek RTL8218B (external)] (irq=POLL)
[   21.227023] rtl83xx-switch switch@1b000000 lan2 (uninitialized): PHY [mdio-bus:01] driver [Realtek RTL8218B (external)] (irq=POLL)
[   21.385187] rtl83xx-switch switch@1b000000 lan3 (uninitialized): PHY [mdio-bus:02] driver [Realtek RTL8218B (external)] (irq=POLL)
[   21.543285] rtl83xx-switch switch@1b000000 lan4 (uninitialized): PHY [mdio-bus:03] driver [Realtek RTL8218B (external)] (irq=POLL)
...
```

change:
```
diff --git a/target/linux/realtek/dts-5.10/rtl8393_panasonic_m48eg-pn28480k.dts b/target/linux/realtek/dts-5.10/rtl8393_panasonic_m48eg-pn28480k.dts
index ca04c271eb..6f44a603f9 100644
--- a/target/linux/realtek/dts-5.10/rtl8393_panasonic_m48eg-pn28480k.dts
+++ b/target/linux/realtek/dts-5.10/rtl8393_panasonic_m48eg-pn28480k.dts
@@ -179,14 +179,6 @@
        #interrupt-cells = <2>;
        interrupt-parent = <&gpio0>;
        interrupts = <18 IRQ_TYPE_EDGE_FALLING>;
-
-       /* GPIO12 (IO1_4): 5x RTL8218B + RTL8218FB */
-       ext_switch_reset {
-               gpio-hog;
-               gpios = <12 GPIO_ACTIVE_HIGH>;
-               output-high;
-               line-name = "ext-switch-reset";
-       };
 };

 &i2c_switch {
@@ -218,6 +210,7 @@
 &ethernet0 {
        mdio-bus {
                compatible = "realtek,rtl838x-mdio";
+               reset-gpios = <&gpio2 12 GPIO_ACTIVE_LOW>;
                #address-cells = <1>;
                #size-cells = <0>;

```

full log: https://gist.github.com/musashino205/372467aa470c4d8c89ced4f95e070499

Reply:
Ideally the kernel would resolve device dependencies from DT, but I suppose that's not a trivial problem. What (probably) happens is that the GPIO layer returns `-EPROBE_DEFER` on the yet-unavailable line, causing the ethernet device to be probed again a bit later. The second time the GPIO device was available, so `reset-gpios` can be acquired as expected.

```
mdio_bus mdio-bus: MDIO device at address 0 is missing.
```
If just device@0 is missing, then maybe the MDIO subsystem is trying to probe the phy too soon after reset. Have you tried playing with the `reset-delay-us` and `reset-post-delay-us` values? Although I wonder why that would cause the ethernet to fail entirely, instead of just the first (eight) port(s).

Reply:
Hm, I tried ```reset-delay-us``` but ```reset-post-delay-us``` was not, I'll try.

Reply:
I tried ```reset-delay-us = <10000>``` and ```reset-post-delay-us = <10000>```, but still broken.

Reply:
That's actually quite bad, because this is the way it's supposed to work. Can you tell what the delay is with `gpio-hog`, between the phy-s being taken out of reset and the driver probing the bus?

Reply:
BTW, the rtl83xx-phy driver doesn't configure RTL8218B phy on RTL839x platform, so the phys may not have to be hard-reset. (on RTL838x, ```rtl8380_configure_ext_rtl8218b``` will be called but not on RTL839x)

Reply:
Ah, that's why this is breaking then. Can you add `reset-gpios` only on the phy-s where it does work? Otherwise you'll have to keep the hog with an explanation of why it needs to be done in this way. I would also suggest you create an issue ticket so this problem (RTL8218B reset on RTL839x) is documented.

Reply:
Hmm, this GPIO pin resets all phys on the MDIO line without internal phy on RTL838x SoC, so can't be used for the specified phys only.
And, RTL8218FB phy on **Switch-M16eG PN28160K** (RTL8382M) and **Switch-M24eG PN28240K** (RTL8382M) will be detected as RTL8218D by the phy driver on RTL838x platform, but RTL8218D won't be configured by the driver. So, ```reset-gpios``` may also not be used on that 2x models. I'll test ```reset-gpios``` on **Switch-M24eG PN28240K**.

Reply:
Oops, I tested on **Switch-M24eG PN28240K**, but in addition to RTL8218FB (detected as RTL8218D), external RTL8218B phy is also broken... (internal RTL8218B phy on RTL8382M SoC is good)
The configuration by the phy driver for RTL8218B may be imcomplete for hard-reset.

So, I would like to keep ```gpio-hog``` on 3x models in this PR.

Reply:
Of course using it on one phy doesn't work if it's physically connected to all phy-s. Silly me :facepalm:  

Then `gpio-hog` with some explanation sounds like an okay way to document this in the meantime.